NoClassDefFound during analysis after java with exportJars = true

```
> last raptor-bloomberg-uploader/compile
[debug] 
[debug] Initial source changes: 
[debug]     removed:Set()
[debug]     added: Set()
[debug]     modified: Set()
[debug] Removed products: Set(E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse$2.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse$ZipInputStreamCallback.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseXmlParser$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\PollingScheduler.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationErrorBuilder.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BatchSubmissionReference.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseGatherer$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversationRegistry.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseGatherer.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\SubmissionReference.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse$3.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseXmlParser.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation$Status$2.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseXmlParser.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationError.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergXmlResponse.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseGatherer$2.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation$Status$1.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation$Status.class, E:\code\raptor\raptor-bloomberg-uploader\target\classes\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseGatherer.class)
[debug] Modified external sources: Set()
[debug] Modified binary dependencies: Set(E:\code\raptor\functional-test\target\raptor-controller-5.0.3-SNAPSHOT.jar)
[debug] Initial directly invalidated sources: Set()
[debug] 
[debug] Sources indirectly invalidated by:
[debug]     product: Set(E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseGatherer.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationErrorBuilder.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseXmlParser.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\SubmissionReference.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseXmlParser.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseGatherer.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\PollingScheduler.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BatchSubmissionReference.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergXmlResponse.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationError.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversationRegistry.java)
[debug]     binary dep: Set(E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java)
[debug]     external source: Set()
[debug] Initially invalidated: Set(E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseGatherer.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationErrorBuilder.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseXmlParser.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\SubmissionReference.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseXmlParser.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseGatherer.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\PollingScheduler.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BatchSubmissionReference.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergXmlResponse.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationError.java, E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversationRegistry.java)
[info] Compiling 17 Java sources to E:\code\raptor\raptor-bloomberg-uploader\target\classes...
[debug] running javac with arguments:
[debug]     -encoding
[debug]     UTF-8
[debug]     -d
[debug]     E:\code\raptor\raptor-bloomberg-uploader\target\classes
[debug]     -cp
[debug]     E:\code\raptor\raptor-bloomberg-uploader\target\classes;E:\code\raptor\test-framework\target\test-framework-5.0.3-SNAPSHOT.jar;E:\code\raptor\database\target\database-5.0.3-SNAPSHOT.jar;E:\code\raptor\raptor-config\target\raptor-config-5.0.3-SNAPSHOT.jar;E:\code\raptor\raptor-common\target\raptor-common-5.0.3-SNAPSHOT.jar;E:\code\raptor\functional-test\target\raptor-controller-5.0.3-SNAPSHOT.jar;
[snip really long classpath]

[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BatchSubmissionReference.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversationRegistry.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseGatherer.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponseXmlParser.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergXmlResponse.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseGatherer.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergResponseXmlParser.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\PollingScheduler.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationError.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\ResponseValidationErrorBuilder.java
[debug]     E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\SubmissionReference.java
[debug] com.sun.tools.javac.Main not found; forking javac instead
[debug] Forking javac: javac @c:/temp/sbt_2c8d01c3/argfile
[debug] javac returned exit code: 0
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.BloombergSpfVersion
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.IsinAndReference
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.StandardReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.BloombergSpfVersion
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.IsinAndReference
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.StandardReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.BloombergSpfVersion
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.NoClassDefFoundError: com/efgfp/raptor/functional/DocumentContextCustomiser
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.DocumentContextCustomiser
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.ManualBloombergXmlFileRepository
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.ReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.BloombergSpfVersion
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.NoClassDefFoundError: com/efgfp/raptor/functional/DocumentContextCustomiser
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.SingleProductReportController
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.IsinAndReference
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.ReportRequestBuilder
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.StandardReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.Product
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.publish.ApprovedOrPendingFileSystemBloombergReportPublisher
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.ReportGroup
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergFileUploader.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.publish.ReportGroupReportPublisher
[warn] Error reading API from class file : java.lang.NoClassDefFoundError: com/efgfp/raptor/functional/DocumentContextCustomiser
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.BloombergSpfVersion
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.IsinAndReference
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.StandardReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.UuidGenerator
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.bloomberg.BloombergSpfVersion
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.IsinAndReference
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\DefaultBloombergSubmissionService.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.StandardReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergResponse.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.ReportGroup
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.IsinAndReference
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.StandardReportType
[warn] Problem processing dependencies of source E:\code\raptor\raptor-bloomberg-uploader\src\main\java\com\efgfp\raptor\functional\bloomberg\submission\BloombergConversation.java : java.lang.ClassNotFoundException: com.efgfp.raptor.functional.publish.ApprovedOrPendingFileSystemBloombergReportPublisher
java.lang.NoClassDefFoundError: com/efgfp/raptor/functional/ReportType
    at java.lang.Class.getDeclaredMethods0(Native Method)
    at java.lang.Class.privateGetDeclaredMethods(Class.java:2427)
    at java.lang.Class.privateGetPublicMethods(Class.java:2547)
    at java.lang.Class.getMethods(Class.java:1410)
    at sbt.ClassToAPI$.structure(ClassToAPI.scala:47)
    at sbt.ClassToAPI$.x$1$1(ClassToAPI.scala:37)
    at sbt.ClassToAPI$.instance$1(ClassToAPI.scala:37)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:38)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:38)
    at xsbti.SafeLazy$Impl._t(SafeLazy.scala:20)
    at xsbti.SafeLazy$Impl.get(SafeLazy.scala:24)
    at xsbti.api.ClassLike.structure(ClassLike.java:27)
    at xsbt.api.TagTypeVariables$$anonfun$tagClass0$1.apply$mcV$sp(TagTypeVariables.scala:45)
    at xsbt.api.TagTypeVariables$$anonfun$tagParameterizedDefinition$1.apply$mcV$sp(TagTypeVariables.scala:66)
    at xsbt.api.TagTypeVariables.scope(TagTypeVariables.scala:140)
    at xsbt.api.TagTypeVariables.tagParameterizedDefinition(TagTypeVariables.scala:64)
    at xsbt.api.TagTypeVariables.tagClass0(TagTypeVariables.scala:43)
    at xsbt.api.TagTypeVariables.tagClass(TagTypeVariables.scala:41)
    at xsbt.api.TagTypeVariables.tagDefinition(TagTypeVariables.scala:34)
    at xsbt.api.TagTypeVariables$$anonfun$tag$1.apply(TagTypeVariables.scala:26)
    at xsbt.api.TagTypeVariables$$anonfun$tag$1.apply(TagTypeVariables.scala:26)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:34)
    at scala.collection.mutable.ArrayOps.foreach(ArrayOps.scala:35)
    at xsbt.api.TagTypeVariables.tag(TagTypeVariables.scala:26)
    at xsbt.api.TagTypeVariables$.apply(TagTypeVariables.scala:12)
    at xsbt.api.SameAPI$.apply(SameAPI.scala:58)
    at sbt.inc.Incremental$.sameSource(Incremental.scala:62)
    at sbt.inc.Incremental$$anonfun$3.apply(Incremental.scala:54)
    at sbt.inc.Incremental$$anonfun$3.apply(Incremental.scala:54)
    at scala.Tuple3$Zipped$$anonfun$filter$1.apply(Tuple3.scala:97)
    at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:123)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.Tuple3$Zipped.filter(Tuple3.scala:93)
    at sbt.inc.Incremental$.changedIncremental(Incremental.scala:54)
    at sbt.inc.Incremental$.cycle(Incremental.scala:37)
    at sbt.inc.Incremental$.compile(Incremental.scala:20)
    at sbt.inc.IncrementalCompile$.apply(Compile.scala:17)
    at sbt.AggressiveCompile.compile2(AggressiveCompile.scala:87)
    at sbt.AggressiveCompile.compile1(AggressiveCompile.scala:38)
    at sbt.AggressiveCompile.apply(AggressiveCompile.scala:28)
    at sbt.Compiler$.apply(Compiler.scala:107)
    at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:420)
    at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:420)
    at sbt.Scoped$$anonfun$hf2$1.apply(Structure.scala:506)
    at sbt.Scoped$$anonfun$hf2$1.apply(Structure.scala:506)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:44)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$9.apply(Structure.scala:304)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$9.apply(Structure.scala:304)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$5.work(System.scala:67)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:221)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:221)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:13)
    at sbt.Execute.work(Execute.scala:227)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:221)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:221)
    at sbt.CompletionService$$anon$1$$anon$2.call(CompletionService.scala:26)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
    at java.lang.Thread.run(Thread.java:619)
Caused by: java.lang.ClassNotFoundException: com.efgfp.raptor.functional.ReportType
    at java.net.URLClassLoader$1.run(URLClassLoader.java:202)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:190)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:307)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:248)
    at java.lang.Class.getDeclaredMethods0(Native Method)
    at java.lang.Class.privateGetDeclaredMethods(Class.java:2427)
    at java.lang.Class.privateGetPublicMethods(Class.java:2547)
    at java.lang.Class.getMethods(Class.java:1410)
    at sbt.ClassToAPI$.structure(ClassToAPI.scala:47)
    at sbt.ClassToAPI$.x$1$1(ClassToAPI.scala:37)
    at sbt.ClassToAPI$.instance$1(ClassToAPI.scala:37)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:38)
    at sbt.ClassToAPI$$anonfun$5.apply(ClassToAPI.scala:38)
    at xsbti.SafeLazy$Impl._t(SafeLazy.scala:20)
    at xsbti.SafeLazy$Impl.get(SafeLazy.scala:24)
    at xsbti.api.ClassLike.structure(ClassLike.java:27)
    at xsbt.api.TagTypeVariables$$anonfun$tagClass0$1.apply$mcV$sp(TagTypeVariables.scala:45)
    at xsbt.api.TagTypeVariables$$anonfun$tagParameterizedDefinition$1.apply$mcV$sp(TagTypeVariables.scala:66)
    at xsbt.api.TagTypeVariables.scope(TagTypeVariables.scala:140)
    at xsbt.api.TagTypeVariables.tagParameterizedDefinition(TagTypeVariables.scala:64)
    at xsbt.api.TagTypeVariables.tagClass0(TagTypeVariables.scala:43)
    at xsbt.api.TagTypeVariables.tagClass(TagTypeVariables.scala:41)
    at xsbt.api.TagTypeVariables.tagDefinition(TagTypeVariables.scala:34)
    at xsbt.api.TagTypeVariables$$anonfun$tag$1.apply(TagTypeVariables.scala:26)
    at xsbt.api.TagTypeVariables$$anonfun$tag$1.apply(TagTypeVariables.scala:26)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:34)
    at scala.collection.mutable.ArrayOps.foreach(ArrayOps.scala:35)
    at xsbt.api.TagTypeVariables.tag(TagTypeVariables.scala:26)
    at xsbt.api.TagTypeVariables$.apply(TagTypeVariables.scala:12)
    at xsbt.api.SameAPI$.apply(SameAPI.scala:58)
    at sbt.inc.Incremental$.sameSource(Incremental.scala:62)
    at sbt.inc.Incremental$$anonfun$3.apply(Incremental.scala:54)
    at sbt.inc.Incremental$$anonfun$3.apply(Incremental.scala:54)
    at scala.Tuple3$Zipped$$anonfun$filter$1.apply(Tuple3.scala:97)
    at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:123)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:322)
    at scala.Tuple3$Zipped.filter(Tuple3.scala:93)
    at sbt.inc.Incremental$.changedIncremental(Incremental.scala:54)
    at sbt.inc.Incremental$.cycle(Incremental.scala:37)
    at sbt.inc.Incremental$.compile(Incremental.scala:20)
    at sbt.inc.IncrementalCompile$.apply(Compile.scala:17)
    at sbt.AggressiveCompile.compile2(AggressiveCompile.scala:87)
    at sbt.AggressiveCompile.compile1(AggressiveCompile.scala:38)
    at sbt.AggressiveCompile.apply(AggressiveCompile.scala:28)
    at sbt.Compiler$.apply(Compiler.scala:107)
    at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:420)
    at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:420)
    at sbt.Scoped$$anonfun$hf2$1.apply(Structure.scala:506)
    at sbt.Scoped$$anonfun$hf2$1.apply(Structure.scala:506)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:44)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$9.apply(Structure.scala:304)
    at sbt.Scoped$Reduced$$anonfun$combine$1$$anonfun$apply$9.apply(Structure.scala:304)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$5.work(System.scala:67)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:221)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:221)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:13)
    at sbt.Execute.work(Execute.scala:227)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:221)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:221)
    at sbt.CompletionService$$anon$1$$anon$2.call(CompletionService.scala:26)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
    at java.lang.Thread.run(Thread.java:619)
[error] {file:/E:/code/raptor/}raptor-bloomberg-uploader/compile: java.lang.NoClassDefFoundError: com/efgfp/raptor/functional/ReportType
> eval "jar tf E:/code/raptor/functional-test/target/raptor-controller-5.0.3-SNAPSHOT.jar" #| "grep com/efgfp/raptor/functional/ReportType" !
com/efgfp/raptor/functional/ReportType.class
[info] ans: Int = 0
> show raptor-controller/export-jars
[info] true
```

Let me know if you need a smaller test case.

--- a/internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
+++ b/internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
@@ -213,7 +213,7 @@ object ClassToAPI {
         c.getConstructors.toIndexedSeq,
         constructorToDef(enclPkg)
       )
-    val classes = innerClassesOf(c, cf, cmap)
+    val classes = innerClassesFromClassfile(c, cf, cmap)
     val all = methods ++ fields ++ constructors ++ classes
     val parentJavaTypes = allSuperTypes(c)
     if (!Modifier.isPrivate(c.getModifiers))
@@ -229,28 +229,7 @@ object ClassToAPI {
     (staticStructure, instanceStructure)
   }
 
-  /** Uses reflection for inner classes, falling back to classfile parsing on error. sbt/sbt#117 */
-  private def innerClassesOf(
-      c: Class[?],
-      cf: => ClassFile,
-      cmap: ClassMap
-  ): Defs =
-    try
-      merge[Class[?]](
-        c,
-        c.getDeclaredClasses.toIndexedSeq,
-        c.getClasses.toIndexedSeq,
-        toDefinitions(cmap),
-        (_: Seq[Class[?]]).partition(isStatic),
-        _.getEnclosingClass != c
-      )
-    catch {
-      case e: (NoClassDefFoundError) =>
-        cmap.log.warn(s"Could not reflect on inner classes of ${c.getName}: $e")
-        cmap.log.warn(s"Falling back to classfile parsing")
-        innerClassesFromClassfile(c, cf, cmap)
-    }
-
+  /** Enumerates inner classes from the classfile instead of reflection. sbt/sbt#117 */
   private def innerClassesFromClassfile(
       c: Class[?],
       cf: => ClassFile,
@@ -259,30 +238,40 @@ object ClassToAPI {
     val cl = c.getClassLoader
     val name = c.getName
     val declaredClasses = new mutable.ArrayBuffer[Class[?]]()
-    val publicClasses = new mutable.ArrayBuffer[Class[?]]()
-    for (info <- cf.innerClasses) {
-      if (info.outerClassName == name) {
-        try {
-          val innerClass = cl.loadClass(info.innerClassName)
-          declaredClasses += innerClass
-          if (info.isPublic) {
-            publicClasses += innerClass
-          }
-        } catch {
-          case e: (ClassNotFoundException | NoClassDefFoundError) =>
-            cmap.log.warn(s"Could not load inner class ${info.innerClassName} of $name: $e")
-        }
-      }
+    val inheritedClasses = new mutable.ArrayBuffer[Class[?]]()
+    // direct inner classes from this class's classfile
+    for (info <- cf.innerClasses if info.outerClassName == name) {
+      loadInnerClass(cl, info, cmap.log).foreach(declaredClasses += _)
+    }
+    // inherited public inner classes from parent classfiles
+    for {
+      parent <- allSuperTypes(c).collect { case c: Class[?] => c }
+      parentCf = classFileForClass(parent)
+      info <- parentCf.innerClasses if info.outerClassName == parent.getName && info.isPublic
+    } {
+      loadInnerClass(cl, info, cmap.log).foreach(inheritedClasses += _)
     }
     merge[Class[?]](
       c,
       declaredClasses.toIndexedSeq,
-      publicClasses.toIndexedSeq,
+      (declaredClasses.filter(cls => Modifier.isPublic(cls.getModifiers)) ++ inheritedClasses).toIndexedSeq,
       toDefinitions(cmap),
       (_: Seq[Class[?]]).partition(isStatic),
       _.getEnclosingClass != c
     )
   }
+
+  private def loadInnerClass(
+      cl: ClassLoader,
+      info: classfile.InnerClassInfo,
+      log: Logger
+  ): Option[Class[?]] =
+    try Some(cl.loadClass(info.innerClassName))
+    catch {
+      case e: (ClassNotFoundException | NoClassDefFoundError) =>
+        log.warn(s"Could not load inner class ${info.innerClassName}: $e")
+        None
+    }
 
   /** TODO: over time, ClassToAPI should switch the majority of access to the classfile parser */
   private def classFileForClass(c: Class[?]): ClassFile =

--- a/internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
+++ b/internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
@@ -28,9 +28,10 @@ object ClassToAPI {
 
   // (api, public inherited classes)
   def process(
-      classes: Seq[Class[?]]
+      classes: Seq[Class[?]],
+      log: Logger = Logger.Null
   ): (Seq[api.ClassLike], Seq[String], Set[(Class[?], Class[?])]) = {
-    val cmap = emptyClassMap
+    val cmap = emptyClassMap(log)
     classes.foreach(toDefinitions(cmap)) // force recording of class definitions
     cmap.lz.toList
       .foreach(_.get()) // force thunks to ensure all inherited dependencies are recorded
@@ -66,7 +67,8 @@ object ClassToAPI {
       private[sbt] val inherited: mutable.Set[(Class[?], Class[?])],
       private[sbt] val lz: mutable.Buffer[xsbti.api.Lazy[?]],
       private[sbt] val allNonLocalClasses: mutable.Set[api.ClassLike],
-      private[sbt] val mainClasses: mutable.Set[String]
+      private[sbt] val mainClasses: mutable.Set[String],
+      private[sbt] val log: Logger
   ) {
     def clear(): Unit = {
       memo.clear()
@@ -74,13 +76,14 @@ object ClassToAPI {
       lz.clear()
     }
   }
-  def emptyClassMap: ClassMap =
+  def emptyClassMap(log: Logger = Logger.Null): ClassMap =
     new ClassMap(
       new mutable.HashMap,
       new mutable.HashSet,
       new mutable.ListBuffer,
       new mutable.HashSet,
-      new mutable.HashSet
+      new mutable.HashSet,
+      log
     )
 
   /**
@@ -210,14 +213,7 @@ object ClassToAPI {
         c.getConstructors.toIndexedSeq,
         constructorToDef(enclPkg)
       )
-    val classes = merge[Class[?]](
-      c,
-      c.getDeclaredClasses.toIndexedSeq,
-      c.getClasses.toIndexedSeq,
-      toDefinitions(cmap),
-      (_: Seq[Class[?]]).partition(isStatic),
-      _.getEnclosingClass != c
-    )
+    val classes = innerClassesOf(c, cf, cmap)
     val all = methods ++ fields ++ constructors ++ classes
     val parentJavaTypes = allSuperTypes(c)
     if (!Modifier.isPrivate(c.getModifiers))
@@ -231,6 +227,61 @@ object ClassToAPI {
       lzyS(all.staticInherited.toArray)
     )
     (staticStructure, instanceStructure)
+  }
+
+  /** Uses reflection for inner classes, falling back to classfile parsing on error. sbt/sbt#117 */
+  private def innerClassesOf(
+      c: Class[?],
+      cf: => ClassFile,
+      cmap: ClassMap
+  ): Defs =
+    try
+      merge[Class[?]](
+        c,
+        c.getDeclaredClasses.toIndexedSeq,
+        c.getClasses.toIndexedSeq,
+        toDefinitions(cmap),
+        (_: Seq[Class[?]]).partition(isStatic),
+        _.getEnclosingClass != c
+      )
+    catch {
+      case e: (NoClassDefFoundError) =>
+        cmap.log.warn(s"Could not reflect on inner classes of ${c.getName}: $e")
+        cmap.log.warn(s"Falling back to classfile parsing")
+        innerClassesFromClassfile(c, cf, cmap)
+    }
+
+  private def innerClassesFromClassfile(
+      c: Class[?],
+      cf: => ClassFile,
+      cmap: ClassMap
+  ): Defs = {
+    val cl = c.getClassLoader
+    val name = c.getName
+    val declaredClasses = new mutable.ArrayBuffer[Class[?]]()
+    val publicClasses = new mutable.ArrayBuffer[Class[?]]()
+    for (info <- cf.innerClasses) {
+      if (info.outerClassName == name) {
+        try {
+          val innerClass = cl.loadClass(info.innerClassName)
+          declaredClasses += innerClass
+          if (info.isPublic) {
+            publicClasses += innerClass
+          }
+        } catch {
+          case e: (ClassNotFoundException | NoClassDefFoundError) =>
+            cmap.log.warn(s"Could not load inner class ${info.innerClassName} of $name: $e")
+        }
+      }
+    }
+    merge[Class[?]](
+      c,
+      declaredClasses.toIndexedSeq,
+      publicClasses.toIndexedSeq,
+      toDefinitions(cmap),
+      (_: Seq[Class[?]]).partition(isStatic),
+      _.getEnclosingClass != c
+    )
   }
 
   /** TODO: over time, ClassToAPI should switch the majority of access to the classfile parser */

--- a/internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassToAPISpecification.scala
+++ b/internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassToAPISpecification.scala
@@ -13,7 +13,9 @@ package sbt
 package internal
 package inc
 
+import java.io.File
 import sbt.internal.inc.classfile.JavaCompilerForUnitTesting
+import sbt.io.IO
 import xsbti.{ AnalysisCallback, VirtualFileRef }
 import xsbti.api.{ ClassLike, ClassLikeDef, DefinitionType }
 
@@ -50,6 +52,52 @@ class ClassToAPISpecification extends UnitSpec {
       """.stripMargin
     val apis = extractApisFromSrc("A.java" -> src).map(c => c.name -> c).toMap
     assert(apis.keySet === Set("A", "A.B"))
+  }
+
+  // sbt/sbt#117
+  it should "not throw NoClassDefFoundError when inner class references missing type" in {
+    IO.withTemporaryDirectory { temp =>
+      val libDir = new File(temp, "lib")
+      val srcDir = new File(temp, "src")
+      libDir.mkdir()
+      srcDir.mkdir()
+
+      val missingFile = new File(temp, "Missing.java")
+      IO.write(missingFile, "public class Missing {}")
+      compileJava(Seq(missingFile), libDir, Seq.empty)
+
+      val outerFile = new File(temp, "Outer.java")
+      IO.write(
+        outerFile,
+        """|public class Outer {
+           |  public class Inner extends Missing {}
+           |  public void hello() {}
+           |}
+           |""".stripMargin
+      )
+      compileJava(Seq(outerFile), srcDir, Seq(libDir))
+
+      val classloader = new java.net.URLClassLoader(Array(srcDir.toURI.toURL), null)
+      val outerClass = classloader.loadClass("Outer")
+      val (apis, _, _) = ClassToAPI.process(Seq(outerClass))
+
+      val names = apis.map(_.name).toSet
+      assert(names.contains("Outer"))
+      assert(!names.contains("Outer.Inner"))
+    }
+  }
+
+  private def compileJava(files: Seq[File], outputDir: File, classpath: Seq[File]): Unit = {
+    import javax.tools.{ StandardLocation, ToolProvider }
+    import scala.jdk.CollectionConverters._
+    val compiler = ToolProvider.getSystemJavaCompiler()
+    val fileManager = compiler.getStandardFileManager(null, null, null)
+    fileManager.setLocation(StandardLocation.CLASS_OUTPUT, Seq(outputDir).asJava)
+    if (classpath.nonEmpty)
+      fileManager.setLocation(StandardLocation.CLASS_PATH, classpath.asJava)
+    val units = fileManager.getJavaFileObjectsFromFiles(files.asJava)
+    compiler.getTask(null, fileManager, null, null, null, units).call()
+    fileManager.close()
   }
 
   /**

--- a/internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/ClassFile.scala
+++ b/internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/ClassFile.scala
@@ -28,6 +28,7 @@ private[sbt] trait ClassFile {
   val methods: Array[FieldOrMethodInfo]
   val attributes: Array[AttributeInfo]
   def sourceFile: Option[String]
+  def innerClasses: Array[InnerClassInfo]
   def types: Set[String]
   def stringValue(a: AttributeInfo): String
 
@@ -95,8 +96,18 @@ private[sbt] final case class AttributeInfo(name: Option[String], value: Array[B
   def isNamed(s: String) = name.exists(s == _)
   def isSignature = isNamed("Signature")
   def isSourceFile = isNamed("SourceFile")
+  def isInnerClasses = isNamed("InnerClasses")
   def isRuntimeVisibleAnnotations = isNamed("RuntimeVisibleAnnotations")
   def isRuntimeInvisibleAnnotations = isNamed("RuntimeInvisibleAnnotations")
+}
+private[sbt] final case class InnerClassInfo(
+    accessFlags: Int,
+    innerName: Option[String],
+    innerClassName: String,
+    outerClassName: String
+) {
+  def isStatic = (accessFlags & ACC_STATIC) == ACC_STATIC
+  def isPublic = (accessFlags & ACC_PUBLIC) == ACC_PUBLIC
 }
 private[sbt] object Constants {
   final val ACC_STATIC = 0x0008

--- a/internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Parser.scala
+++ b/internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Parser.scala
@@ -17,7 +17,7 @@ package classfile
 import java.net.URL
 import java.nio.file.Files
 import java.nio.file.Path
-import java.io.{ BufferedInputStream, InputStream, File, DataInputStream }
+import java.io.{ BufferedInputStream, ByteArrayInputStream, InputStream, File, DataInputStream }
 
 import sbt.internal.io.ErrorHandling
 
@@ -79,6 +79,33 @@ private[sbt] object Parser {
       val methods = readFieldsOrMethods()
 
       val attributes = array(in.readUnsignedShort())(parseAttribute())
+
+      lazy val innerClasses: Array[InnerClassInfo] = {
+        attributes
+          .find(_.isInnerClasses)
+          .map(parseInnerClasses)
+          .getOrElse(Array.empty)
+      }
+
+      private def parseInnerClasses(a: AttributeInfo): Array[InnerClassInfo] = {
+        val bais = new ByteArrayInputStream(a.value)
+        val data = new DataInputStream(bais)
+        val numberOfClasses = data.readUnsignedShort()
+        Array.tabulate(numberOfClasses) { _ =>
+          val innerClassInfoIndex = data.readUnsignedShort()
+          val outerClassInfoIndex = data.readUnsignedShort()
+          val innerNameIndex = data.readUnsignedShort()
+          val innerClassAccessFlags = data.readUnsignedShort()
+          val innerName = if (innerNameIndex == 0) None else Some(toUTF8(innerNameIndex))
+          val innerCN =
+            if (innerClassInfoIndex == 0) ""
+            else getClassConstantName(innerClassInfoIndex)
+          val outerCN =
+            if (outerClassInfoIndex == 0) ""
+            else getClassConstantName(outerClassInfoIndex)
+          InnerClassInfo(innerClassAccessFlags, innerName, innerCN, outerCN)
+        }
+      }
 
       override lazy val sourceFile: Option[String] =
         for (sourceFileAttribute <- attributes.find(_.isSourceFile))

--- a/internal/zinc-classfile/src/test/scala/sbt/internal/inc/classfile/ParserSpecification.scala
+++ b/internal/zinc-classfile/src/test/scala/sbt/internal/inc/classfile/ParserSpecification.scala
@@ -42,4 +42,26 @@ class ParserSpecification extends UnitSpec {
       assert(classfile.types.nonEmpty)
     }
 
+  it should "parse InnerClasses attribute for AbstractMap.SimpleEntry" in {
+    val logger = ConsoleLogger()
+    val c = classOf[java.util.AbstractMap.SimpleEntry[String, String]]
+    val cf = Parser(sbt.io.IO.classfileLocation(c), logger)
+    val innerClasses = cf.innerClasses
+    assert(innerClasses.nonEmpty)
+    val self = innerClasses.find(_.innerClassName == "java.util.AbstractMap$SimpleEntry")
+    assert(self.isDefined)
+    assert(self.get.outerClassName == "java.util.AbstractMap")
+  }
+
+  it should "parse InnerClasses attribute for AbstractMap" in {
+    val logger = ConsoleLogger()
+    val c = classOf[java.util.AbstractMap[?, ?]]
+    val cf = Parser(sbt.io.IO.classfileLocation(c), logger)
+    val innerClasses = cf.innerClasses
+    val entry = innerClasses.find(_.innerClassName == "java.util.AbstractMap$SimpleEntry")
+    assert(entry.isDefined)
+    assert(entry.get.outerClassName == "java.util.AbstractMap")
+    assert(entry.get.isPublic)
+  }
+
 }

--- a/zinc/src/main/scala/sbt/internal/inc/javac/AnalyzingJavaCompiler.scala
+++ b/zinc/src/main/scala/sbt/internal/inc/javac/AnalyzingJavaCompiler.scala
@@ -184,7 +184,7 @@ final class AnalyzingJavaCompiler private[sbt] (
 
       // Read the API information from [[Class]] to analyze dependencies.
       def readAPI(source: VirtualFileRef, classes: Seq[Class[?]]): Set[(String, String)] = {
-        val (apis, mainClasses, inherits) = ClassToAPI.process(classes)
+        val (apis, mainClasses, inherits) = ClassToAPI.process(classes, log)
         apis.foreach(callback.api(source, _))
         mainClasses.foreach(callback.mainClass(source, _))
         inherits.map {

